### PR TITLE
🐛 Fix disappearing task names in task history

### PR DESCRIPTION
**Pull request by kiwina**
### Related GitHub Issue

Closes: #4410

### Description

This PR fixes the issue where task history entries appear with blank or missing names, making it impossible for users to identify and resume previous tasks. The core problem was that the `getTaskMetadata` function didn't handle cases where tasks had no messages or empty/undefined names.

**Key implementation details:**
- Added robust fallback logic in `taskMetadata.ts` to ensure tasks always have meaningful names
- Implemented crash-resistant task naming that works even when tasks are incomplete or interrupted
- Enhanced error handling for edge cases where message arrays are empty or malformed
- Ensured backward compatibility with existing task history entries

**Specific changes:**
- Modified `getTaskMetadata()` to provide fallback names like "Untitled Task", "Incomplete Task", or "Task from [timestamp]"
- Added comprehensive error handling for tasks with no messages
- Fixed TypeScript type issues in the return value structure
- Improved task persistence reliability during application crashes or interruptions

### Test Procedure

**Manual Testing Steps:**
1. Install the updated extension
2. Create a new task and let it run partially, then force-close VS Code
3. Restart VS Code and check the task history - the task should have a meaningful name
4. Create a task with no initial message and verify it gets a fallback name
5. Test with various edge cases (empty messages, undefined names, etc.)

**Verification:**
- Task history entries now consistently show meaningful names
- No more blank entries in the task list
- Previously created tasks with missing names are handled gracefully
- Application remains stable during task interruptions

### Type of Change

- [x] 🐛 **Bug Fix**: Non-breaking change that fixes an issue.

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Code Quality**: My code adheres to the project's style guidelines.
- [x] **Testing**: The application builds successfully with my changes.
- [x] **Branch Hygiene**: My branch is up-to-date (rebased) with the `main` branch.
- [x] **Contribution Guidelines**: I have read and agree to the Contributor Guidelines.

### Documentation Updates

- [x] No documentation updates are required.

### Additional Notes

This fix addresses a critical UX issue where users couldn't identify their previous tasks due to missing names. The solution is backward-compatible and handles various edge cases including:

- Tasks interrupted by application crashes
- Tasks with no initial messages
- Tasks with empty or undefined names
- Malformed task data

The implementation prioritizes data safety and ensures that even in failure scenarios, users can still access and identify their task history.

### Get in Touch

Available for questions and feedback on this PR.

----

> [!IMPORTANT]
> Fixes disappearing task names in `taskMetadata.ts` by adding fallback logic and improving error handling for edge cases.
> 
>   - **Behavior**:
>     - Fixes disappearing task names in `taskMetadata()` by adding fallback names like "Untitled Task" or "Incomplete Task".
>     - Handles cases with no messages by assigning a default name "Task #[number] (No messages)".
>     - Ensures task names are never blank, even if tasks are incomplete or interrupted.
>   - **Error Handling**:
>     - Adds error handling for empty or malformed message arrays in `taskMetadata.ts`.
>     - Fixes TypeScript type issues in the return value structure of `taskMetadata()`.
>   - **Misc**:
>     - Improves task persistence reliability during application crashes or interruptions.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for d93cd9250ad600dff022dd80ad9e86b860640ca5. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/core/task-persistence/taskMetadata.ts [hunk 1]
```diff
@@ -27,10 +27,40 @@ export async function taskMetadata({
 	workspace,
 }: TaskMetadataOptions) {
 	const taskDir = await getTaskDirectoryPath(globalStoragePath, taskId)
+
+	// Handle edge case where there are no messages at all
+	if (!messages || messages.length === 0) {
```
Comment (daniel-lxs) [inline src/core/task-persistence/taskMetadata.ts:32]: Consider refactoring to eliminate code duplication here. The `historyItem` object is created twice - once in this early return and once at the end of the function. You could restructure this by:

1. Using a boolean flag like `const hasMessages = messages && messages.length > 0`
2. Pre-calculating all values (`taskName`, `timestamp`, `tokenUsage`, `taskDirSize`) based on message availability
3. Creating the `historyItem` object only once at the end

This would improve maintainability and follow DRY principles.

File: src/core/task-persistence/taskMetadata.ts [hunk 2]
```diff
@@ -27,10 +27,40 @@ export async function taskMetadata({
...
+			},
+		}
+	}
+
 	const taskMessage = messages[0] // First message is always the task say.
 
 	const lastRelevantMessage =
-		messages[findLastIndex(messages, (m) => !(m.ask === "resume_task" || m.ask === "resume_completed_task"))]
+		messages[findLastIndex(messages, (m) => !(m.ask === "resume_task" || m.ask === "resume_completed_task"))] ||
+		messages[0]
```
Comment (daniel-lxs) [inline src/core/task-persistence/taskMetadata.ts:64]: The `tokenUsage` object is missing the `contextTokens` field. Based on the `TokenUsage` type definition, it should include:

```typescript
tokenUsage: {
    totalTokensIn: 0,
    totalTokensOut: 0,
    totalCacheWrites: 0,
    totalCacheReads: 0,
    totalCost: 0,
    contextTokens: 0,  // <- Missing field
}
```

This ensures type consistency with the rest of the codebase.

File: src/core/task-persistence/taskMetadata.ts [hunk 3]
```diff
@@ -45,11 +75,17 @@ export async function taskMetadata({
 
 	const tokenUsage = getApiMetrics(combineApiRequests(combineCommandSequences(messages.slice(1))))
 
+	// Ensure task name is never blank - provide fallback names
+	let taskName = taskMessage.text?.trim() || ""
+	if (!taskName) {
+		taskName = `Task #${taskNumber} (Incomplete)`
+	}
+
 	const historyItem: HistoryItem = {
```
Comment (daniel-lxs) [inline src/core/task-persistence/taskMetadata.ts:84]: Would it be better to restructure the function to use a single object creation pattern:

```typescript
// Determine message availability upfront
const hasMessages = messages && messages.length > 0

// Pre-calculate all values based on availability
let taskName: string
let timestamp: number
let tokenUsage: ReturnType<typeof getApiMetrics>
let taskDirSize: number

if (!hasMessages) {
    // Handle no messages case
} else {
    // Handle messages case
}

// Create historyItem once with pre-calculated values
```

This eliminates the duplicate object creation and makes the code more maintainable.

File: src/core/task-persistence/taskMetadata.ts [hunk 4]
```diff
@@ -27,10 +27,40 @@ export async function taskMetadata({
...
+			},
+		}
+	}
+
 	const taskMessage = messages[0] // First message is always the task say.
 
 	const lastRelevantMessage =
-		messages[findLastIndex(messages, (m) => !(m.ask === "resume_task" || m.ask === "resume_completed_task"))]
+		messages[findLastIndex(messages, (m) => !(m.ask === "resume_task" || m.ask === "resume_completed_task"))] ||
+		messages[0]
```
Comment (daniel-lxs) [inline src/core/task-persistence/taskMetadata.ts:63]: It should be possible to use `taskMessage` here:
```suggestion
		taskMessage
```

File: src/core/task-persistence/taskMetadata.ts [hunk 5]
```diff
@@ -45,11 +75,17 @@ export async function taskMetadata({
...
+	if (!taskName) {
+		taskName = `Task #${taskNumber} (Incomplete)`
+	}
+
 	const historyItem: HistoryItem = {
 		id: taskId,
 		number: taskNumber,
 		ts: lastRelevantMessage.ts,
-		task: taskMessage.text ?? "",
+		task: taskName,
```
Comment (daniel-lxs) [inline src/core/task-persistence/taskMetadata.ts:88]: Could this work better here? That way `taskName` is not reassigned unnecessarily.  

```suggestion
		task: taskMessage.text?.trim() || `Task #${taskNumber} (Incomplete)`,
```

**Comment by daniel-lxs** [comment]
Thank you for your contribution! 
Since I couldn't push the suggested changes I created a new PR https://github.com/RooCodeInc/Roo-Code/pull/5071
